tests/subsys/power/power_mgmt_soc/src/power_mgmt.c

```diff
@@ -0,0 +1,326 @@
+/*
+ * Copyright (c) 2020 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <zephyr.h>
+#include <device.h>
+#include <soc.h>
+#include <power/power.h>
+#include <logging/log.h>
+#define LOG_LEVEL LOG_LEVEL_DBG
+LOG_MODULE_REGISTER(pwrmgmt_test);
+
+#define SLP_STATES_SUPPORTED      2ul
+
+/* Thread properties */
+#define TASK_STACK_SIZE           1024ul
+#define PRIORITY                  K_PRIO_COOP(5)
+
+/* Sleep time should be lower than CONFIG_SYS_PM_MIN_RESIDENCY_SLEEP_1 */
+#define THREAD_A_SLEEP_TIME       100ul
+#define THREAD_B_SLEEP_TIME       1000ul
+
+/* Maximum latency should be less than 300 ms */
+#define MAX_EXPECTED_MS_LATENCY   500ul
+
+/* Sleep some extra time than minimum residency */
+#define DP_EXTRA_SLP_TIME         1100ul
+#define LT_EXTRA_SLP_TIME         500ul
+
+#define SEC_TO_MSEC               1000ul
+
+K_THREAD_STACK_DEFINE(stack_a, TASK_STACK_SIZE);
+K_THREAD_STACK_DEFINE(stack_b, TASK_STACK_SIZE);
+
+static struct k_thread thread_a_id;
+static struct k_thread thread_b_id;
+
+struct pm_counter {
+	uint8_t entry_cnt;
+	uint8_t exit_cnt;
+};
+
+/* Track time elapsed */
+static int64_t trigger_time;
+static bool checks_enabled;
+/* Track entry/exit to sleep */
+struct pm_counter pm_counters[SLP_STATES_SUPPORTED];
+
+static void pm_latency_check(void)
+{
+	int64_t latency;
+	int secs;
+	int msecs;
+
+	latency = k_uptime_delta(&trigger_time);
+	secs = (int)(latency / SEC_TO_MSEC);
+	msecs = (int)(latency % SEC_TO_MSEC);
+	LOG_INF("PM sleep entry latency %d.%03d seconds", secs, msecs);
+
+	if (secs > 0) {
+		LOG_WRN("Sleep entry latency is too high");
+		return;
+	}
+
+	if (msecs > MAX_EXPECTED_MS_LATENCY) {
+		LOG_WRN("Sleep entry latency is higher than expected");
+	}
+}
+
+/* Hooks to count entry/exit */
+void sys_pm_notify_power_state_entry(enum power_states state)
+{
+	if (!checks_enabled) {
+		return;
+	}
+
+	switch (state) {
+	case SYS_POWER_STATE_SLEEP_1:
+		pm_counters[0].entry_cnt++;
+		pm_latency_check();
+		break;
+	case SYS_POWER_STATE_DEEP_SLEEP_1:
+		pm_counters[1].entry_cnt++;
+		pm_latency_check();
+		break;
+	default:
+		break;
+	}
+}
+
+void sys_pm_notify_power_state_exit(enum power_states state)
+{
+	if (!checks_enabled) {
+		return;
+	}
+
+	switch (state) {
+	case SYS_POWER_STATE_SLEEP_1:
+		pm_counters[0].exit_cnt++;
+		break;
+	case SYS_POWER_STATE_DEEP_SLEEP_1:
+		pm_counters[1].exit_cnt++;
+		break;
+	default:
+		break;
+	}
+}
+
+static void pm_check_counters(uint8_t cycles)
+{
+	for (int i = 0; i < SLP_STATES_SUPPORTED; i++) {
+		LOG_INF("PM state[%d] entry counter %d\n", i,
+			pm_counters[i].entry_cnt);
+		LOG_INF("PM state[%d] exit counter %d\n", i,
+			pm_counters[i].exit_cnt);
+
+		if (pm_counters[i].entry_cnt != pm_counters[i].exit_cnt) {
+			LOG_WRN("PM counters entry/exit mismatch");
+		}
+
+		if (pm_counters[i].entry_cnt != cycles) {
+			LOG_WRN("PM counter mismatch expected: %d", cycles);
```

Comment:
I get this on `cc1352r1_launchxl`:
```
I: PM sleep residency 3.104 seconds                                             
I: Resume                                                                       
ABI: PM multi-thread completed                                                  
I: PM state[0] entry counter 0                                                  
                                                                                
I: PM state[0] exit counter 0                                                   
                                                                                
W: PM counter mismatch expected: 5                                              
I: PM state[1] entry counter 0                                                  
                                                                                
I: PM state[1] exit counter 0                                                   
                                                                                
W: PM counter mismatch expected: 5                                              
Power management test finished.                                                 
 PASS - main                                                                    
===================================================================             
===================================================================             
PROJECT EXECUTION SUCCESSFUL     
```
The test passes, but there are warning-level log messages that a value isn't what is expected.

Either the test should not be passing, or those messages should be downgraded to INF.

Reply:
cc1352 is not supposed to enter deep sleep state. I've changed the if statement at to filter out boards that support sleep only. Can you check the latest commit once again and see if cc1352 would still enter deep sleep? Thanks.

Reply:
No, this isn't working.

<details>
  <summary>Click to expand trace of cc1352r1_launchxl output</summary>

```
*** Booting Zephyr OS build zephyr-v2.3.0-1348-g118693b273ce  ***
Start power management test.
I: PM single-thread test started for cycles: 3
I: About to enter light sleep
PM >
I: PM sleep entry latency 0.000 seconds
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: About to enter deep Sleep
PM >
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.104 seconds
I: About to enter light sleep
PM >
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: About to enter deep Sleep
PM >
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.104 seconds
I: About to enter light sleep
PM >
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: About to enter deep Sleep
PM >
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.104 seconds
I: PM single-thread completed
I: PM state[0] entry counter 1

I: PM state[0] exit counter 1

W: PM counter mismatch expected: 3
I: PM state[1] entry counter 0

I: PM state[1] exit counter 0

W: PM counter mismatch expected: 3
I: Thread task A init
I: Thread task B init
W: PM multi-thread test started for cycles: 3
I: Suspend...
I: About to enter light sleep
PM >
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: Resume
ABI: Suspend...
I: About to enter deep sleep
PM >
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.103 seconds
I: Resume
ABI: Suspend...
I: About to enter light sleep
PM >
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: Resume
ABI: Suspend...
I: About to enter deep sleep
PM >
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.103 seconds
I: Resume
ABI: Suspend...
I: About to enter light sleep
PM >
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.504 seconds
I: Resume
ABI: Suspend...
I: About to enter deep sleep
PM >
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.104 seconds
I: Resume
ABI: PM multi-thread completed
I: PM state[0] entry counter 0

I: PM state[0] exit counter 0

W: PM counter mismatch expected: 3
I: PM state[1] entry counter 0

I: PM state[1] exit counter 0

W: PM counter mismatch expected: 3
Power management test finished.
 PASS - main
===================================================================
===================================================================
PROJECT EXECUTION SUCCESSFUL
```
</details>

It still says it's trying to enter deep sleep; it still enters deep sleep.

Also if those warnings indicate the test doesn't pass they need to be asserts that cause the test to fail.

Reply:
Can you check the latest commit again and show me the output? Thanks.

Reply:
Doesn't seem a lot different.  You haven't changed the warnings to assertion checks.

<details><summary>Expand for log</summary>

```
 *** Booting Zephyr OS build zephyr-v2.3.0-1348-g71ad5d8ee869  ***
Start power management test.
I: PM single-thread test started for cycles: 2
I: About to enter light sleep
PM >
I: PM sleep entry latency 0.000 seconds
I: PM sleep entry latency 0.043 seconds
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.460 seconds
I: About to enter deep Sleep
PM >
I: PM sleep entry latency 0.001 seconds
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.103 seconds
I: About to enter light sleep
PM >
I: PM sleep entry latency 0.000 seconds
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: About to enter deep Sleep
PM >
I: PM sleep entry latency 0.001 seconds
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.103 seconds
I: PM single-thread completed
I: PM state[0] entry counter 5

I: PM state[0] exit counter 5

W: PM counter mismatch expected: 2
I: PM state[1] entry counter 0

I: PM state[1] exit counter 0

W: PM counter mismatch expected: 2
I: Thread task A init
I: Thread task B init
W: PM multi-thread test started for cycles: 2
I: Suspend...
I: About to enter light sleep
PM >
I: PM sleep entry latency 0.000 seconds
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: Resume
ABI: Suspend...
I: About to enter deep sleep
PM >
I: PM sleep entry latency 0.000 seconds
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.103 seconds
I: Resume
ABI: Suspend...
I: About to enter light sleep
PM >
I: PM sleep entry latency 0.001 seconds
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: Resume
ABI: Suspend...
I: About to enter deep sleep
PM >
I: PM sleep entry latency 0.001 seconds
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.103 seconds
I: Resume
ABI: PM multi-thread completed
I: PM state[0] entry counter 4

I: PM state[0] exit counter 4

W: PM counter mismatch expected: 2
I: PM state[1] entry counter 0

I: PM state[1] exit counter 0

W: PM counter mismatch expected: 2
Power management test finished.
 PASS - main
===================================================================
===================================================================
PROJECT EXECUTION SUCCESSFUL
```
</details>

Reply:
Yea right now I want to make sure the number it enters and exits sleep/deep sleep is correct. The warning there is just comparing the number of entry/exit with total iterations, but more importantly the number of entry/exit counters should match. I will change the warnings to assertions later. For now I need to figure out why this board can't enter deep sleep but I can only see the output with the help from you.

Reply:
Can you paste the log for the latest commit please? Thank you again.

Reply:
If there's a change it isn't obvious to me.  Still says it passed, and still prints warnings about counter mismatch from expected.

<details><summary>Expand for log</summary>

*** Booting Zephyr OS build zephyr-v2.3.0-1348-g35b2cda735e7  ***
Start power management test.
I: PM single-thread test started for cycles: 2
I: About to enter light sleep
PM >
I: PM sleep entry latency 0.000 seconds
I: PM sleep entry latency 0.043 seconds
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.460 seconds
I: About to enter deep Sleep
PM >
I: PM sleep entry latency 0.001 seconds
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.103 seconds
I: About to enter light sleep
PM >
I: PM sleep entry latency 0.000 seconds
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: About to enter deep Sleep
PM >
I: PM sleep entry latency 0.001 seconds
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.103 seconds
I: PM single-thread completed
I: PM state[0] entry counter 5

I: PM state[0] exit counter 5

W: PM counter mismatch expected: 2
I: PM state[1] entry counter 0

I: PM state[1] exit counter 0

W: PM counter mismatch expected: 2
I: Thread task A init
I: Thread task B init
W: PM multi-thread test started for cycles: 2
I: Suspend...
I: About to enter light sleep
PM >
I: PM sleep entry latency 0.000 seconds
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: Resume
ABI: Suspend...
I: About to enter deep sleep
PM >
I: PM sleep entry latency 0.000 seconds
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.103 seconds
I: Resume
ABI: Suspend...
I: About to enter light sleep
PM >
I: PM sleep entry latency 0.001 seconds
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: Resume
ABI: Suspend...
I: About to enter deep sleep
PM >
I: PM sleep entry latency 0.001 seconds
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.103 seconds
I: Resume
ABI: PM multi-thread completed
I: PM state[0] entry counter 4

I: PM state[0] exit counter 4

W: PM counter mismatch expected: 2
I: PM state[1] entry counter 0

I: PM state[1] exit counter 0

W: PM counter mismatch expected: 2
Power management test finished.
 PASS - main
===================================================================
===================================================================
PROJECT EXECUTION SUCCESSFUL
</details>

Reply:
It should be fixed with the latest commit. Only mismatches between entry counter and exit counter are evaluated now. In addition, I believe the cc1352 cannot enter deep sleep as expected because the low-level implementation for power management on the board is not correct. The test case itself is straightforward and the code logic is all correct. I don't think this should be the reason to block the merge for now. The board is supposed to support deep sleep. It enters and exits light sleep correctly but not deep sleep. But the functions of the test case are already good for both micro chips and cc1352, cc26x2.

Reply:
@vanti please check this.  (I should have brought you in before, sorry, thought you were already on copy.).

Reply:
@pabigot Thanks for looping me in.
 
I ran the example on the cc1352, and did not see any warnings (most likely they have been fixed given the code seems to have gone through several iterations). To clarify, the cc1352 does support one deep sleep state.

Looking at the test, I see it expects deep sleep to wake up and resume from the same point of execution as where it went to sleep. That is unexpected - because to my understanding, deep sleep is not required to resume from the same point. The current definition of deep sleep in Zephyr documentation is as follows:

"CPU is power gated and loses execution context. Execution will resume at OS startup code or at a resume point determined by a bootloader that supports deep sleep resume."

On the cc1352, the deep sleep state corresponds to a shutdown state and resumes from the reset vector, hence by design the residency policy is not setup to automatically enter the deep sleep state. It wouldn't make sense for it to do so. Instead it should be manually force-triggered via sys_pm_force_power_state(). See samples/boards/ti/cc13x2_cc26x2/system_off to see how it is handled.

If we want this test to be truly generic, it should not make the assumption that the code execution would resume from the point where the CPU went into deep sleep. The way it is coded currently would make the cc1352 enter light sleep instead, which is indeed what I am seeing. I would suggest adding a comment on lines 253 and 258 of power_mgmt.c stating that platforms that do not automatically enter deep sleep states in its residency policy would simply enter a sleep state.

Reply:
@vanti Thank you for the valuable input! I've added comments as suggested. In addition, changed to ztest framework, including adding assertions at the entry/exit counters' comparison.

Reply:
@pabigot Can you check the latest commit? Right now it should be able to fail if there's any counter mismatch.